### PR TITLE
Skip reader for arbitrary thrift binary protocol

### PR DESCRIPTION
**Pull request by kriskowal**
- [x] tests that exercise all types. could potentially borrow from existing valid byte arrays for other tests.

cc @malandrew

### CODE REVIEW

**Comment by kriskowal** [comment]
This is good to go.

r @malandrew @leizha @jcorbin

**Comment by Raynos** [comment]
What is the purpose and point of a skip reader ?

**Comment by kriskowal** [comment]
@raynos The upcoming static struct RW only reads the fields on the wire that it recognizes and validates their types. If it encounters a type that it does not recognize, it needs to skip it, using the type information on the wire to scan to its end, but does not need to create instances of the resulting value or object graph. The skipField function makes this possible.

**Comment by Raynos** [comment]
@kriskowal cool :=1:

File: skip.js [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+            what: 'field::type'
+        }));
+    }
+
+    return new ReadResult(null, offset);
+}
+
+function skipStruct(buffer, offset) {
+    var result;
+    for (;;) {
```
Comment (jcorbin) [inline skip.js]: suggest a `while (offset < buffer.length)` with a fall-through error return

File: skip.js [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            return new ReadResult(ShortBufferError({
+                expected: offset + 2,
+                actual: buffer.length,
+                buffer: buffer,
+                offset: offset
+            }));
+        }
+        offset += 2;
+
+        result = skipType(buffer, offset, typeid);
```
Comment (jcorbin) [inline skip.js]: suggest push down `var result` to here to avoid accidental (lint obvious) erroneous usage.

**Pull request merged**